python/ray/tests/test_memory_deadlock.py

```diff
@@ -0,0 +1,190 @@
+    ) as _:
+        bytes1 = get_additional_bytes_to_reach_memory_usage_pct(
+            memory_usage_threshold_fraction - 0.1
+        )
+        bytes2 = (
+            get_additional_bytes_to_reach_memory_usage_pct(
+                memory_usage_threshold_fraction + 0.2
+            )
+            - bytes1
+        )
+        ray.get(task_with_nested_task.remote(bytes1, bytes2, None, None))
+
+
+@ray.remote
+def task_with_nested_actor(fraction1, fraction2):
+    leaker = Leaker.options(max_restarts=1, max_task_retries=1).remote()
+    actor_bytes = get_additional_bytes_to_reach_memory_usage_pct(fraction1)
+    ray.get(leaker.allocate.remote(actor_bytes))
+
+    parent_bytes = get_additional_bytes_to_reach_memory_usage_pct(fraction2)
+    dummy = alloc_mem(parent_bytes)
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_task_with_nested_actor(
+    ray_with_memory_monitor,
+):
+    with pytest.raises(
+        (ray.exceptions.OutOfMemoryError, ray.exceptions.RayTaskError)
+    ) as _:
+        ray.get(task_with_nested_actor.remote(0.7, 0.9))
+
+
+@ray.remote
+class ActorWithNestedTask:
+    def __init__(self):
+        self.mem = []
+
+    def alloc_local(self, num_bytes):
+        self.mem = alloc_mem(num_bytes)
+
+    def alloc_remote(self, num_bytes):
+        ray.get(allocate_memory.options(max_retries=1).remote(num_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_actor_with_nested_task(
+    ray_with_memory_monitor,
+):
+    leaker = ActorWithNestedTask.options(max_restarts=1, max_task_retries=1).remote()
+    bytes_first = get_additional_bytes_to_reach_memory_usage_pct(0.45)
+    ray.get(leaker.alloc_local.remote(bytes_first))
+
+    bytes_second = get_additional_bytes_to_reach_memory_usage_pct(0.9)
+    with pytest.raises(ray.exceptions.RayTaskError) as _:
+        ray.get(leaker.alloc_remote.remote(bytes_second))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_actor_with_nested_task_two(
+    ray_with_memory_monitor,
+):
+    leaker1 = ActorWithNestedTask.options(max_restarts=1, max_task_retries=1).remote()
+    leaker2 = ActorWithNestedTask.options(max_restarts=1, max_task_retries=1).remote()
+    bytes_first = get_additional_bytes_to_reach_memory_usage_pct(0.3)
+    ray.get(leaker1.alloc_local.remote(bytes_first))
+    bytes_second = get_additional_bytes_to_reach_memory_usage_pct(0.6)
+    ray.get(leaker2.alloc_local.remote(bytes_second))
+
+    bytes_third = get_additional_bytes_to_reach_memory_usage_pct(0.9)
+    with pytest.raises(ray.exceptions.RayTaskError) as _:
+        ray.get(leaker1.alloc_remote.remote(bytes_third))
+
+
+# Used for syncing allocations
+@ray.remote
+class GlobalActor:
```

Comment:
GlobalActor -> BarrierActor

Reply:
can we make this a threaded actor? https://docs.ray.io/en/latest/ray-core/actors/async_api.html#threaded-actors

Also, can we use threading.Barrier and make all_done a blocking call?